core/src/main/java/com/orientechnologies/orient/core/record/impl/ODocument.java

```diff
@@ -3169,4 +3366,341 @@ protected void unTrack(OIdentifiable id) {
     if (isTrackingChanges() && id.getIdentity().getClusterId() != -2)
       super.unTrack(id);
   }
+
+  private class TypeValue {
+    Object               value;
+    UpdateDeltaValueType type;
+  }
+
+  //this method is callen only for changed fields
+  private TypeValue getUpdateDeltaValue(Object currentValue, Object previousValue, boolean changed, ODocumentEntry parent,
+      List<Object> ownersTrace) {
+    //TODO review this clause once again
+    ownersTrace.add(currentValue);
+    if (changed || !(currentValue instanceof ODocument)) {
+      OType currentType = OType.getTypeByClass(currentValue.getClass());
+      TypeValue retVal = new TypeValue();
+      //separate processing for embedded lists
+      //if previous value is null we want to send whole list to reduce overhead
+      if (currentType.isList() && previousValue != null) {
+        OType previousType = OType.getTypeByClass(previousValue.getClass());
+        //if previous value differs in type we want to send whole list to reduce overhead
+        if (previousType.isList()) {
+          retVal.type = UpdateDeltaValueType.LIST_UPDATE;
+          List deltaList = new ArrayList();
+          retVal.value = deltaList;
+          //now find a diff
+          List currentList = (List) currentValue;
+          List previousList = (List) previousValue;
+          int i = 0;
+          for (i = 0; i < currentList.size(); i++) {
+            Object currentElement = currentList.get(i);
+            if (i > previousList.size() - 1) {
+              //these elements doesn't have pair in original list
+              ODocumentDelta deltaElement = new ODocumentDelta();
+              deltaElement.field("t", UpdateDeltaValueType.LIST_ELEMENT_ADD.getOrd());

```

Comment:
Use a map to set values in the ODocumentDelta is definitely fine, anyway if we have this fields coded in our code would be more efficient have the real java fields inside the ODocumentDelta, anyway this is not blocking i'm happy to merge this part as it is now, and optimize it in a later future.